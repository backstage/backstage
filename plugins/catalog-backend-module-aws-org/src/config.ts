/*
 * Copyright 2024 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  readSchedulerServiceTaskScheduleDefinitionFromConfig,
  SchedulerServiceTaskScheduleDefinition,
} from '@backstage/backend-plugin-api';
import { Config } from '@backstage/config';

export type AwsOrganizationEntityProviderConfig = {
  id: string;
  accountId: string;
  schedule?: SchedulerServiceTaskScheduleDefinition;
};

export function readProviderConfigs(
  config: Config,
): AwsOrganizationEntityProviderConfig[] {
  const providersConfig = config.getOptionalConfig('catalog.providers.aws-org');
  if (!providersConfig) {
    return [];
  }

  return providersConfig.keys().map(id => {
    const providerConfig = providersConfig.getConfig(id);

    return readProviderConfig(id, providerConfig);
  });
}

function readProviderConfig(
  id: string,
  config: Config,
): AwsOrganizationEntityProviderConfig {
  const accountId = config.getString('accountId');

  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;

  return {
    id,
    accountId,
    schedule,
  };
}
