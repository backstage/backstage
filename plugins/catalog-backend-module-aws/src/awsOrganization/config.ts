/*
 * Copyright 2020 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Config } from '@backstage/config';

/**
 * The configuration parameters for a single AWS Organization Processor
 */
export type AwsOrganizationProviderConfig = {
  /**
   * The role to assume for the processor.
   * @deprecated Use `accountId` instead.
   */
  roleArn?: string;

  /**
   * The AWS accountId
   */
  accountId?: string;
};

export function readAwsOrganizationConfig(
  config: Config,
): AwsOrganizationProviderConfig {
  const providerConfig = config.getOptionalConfig('provider');

  const roleArn = providerConfig?.getOptionalString('roleArn');
  const accountId = providerConfig?.getOptionalString('accountId');
  return {
    roleArn,
    accountId,
  };
}
