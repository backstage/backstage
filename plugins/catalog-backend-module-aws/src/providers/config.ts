/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Config } from '@backstage/config';
import { AwsS3Config } from './types';
import { readSchedulerServiceTaskScheduleDefinitionFromConfig } from '@backstage/backend-plugin-api';

const DEFAULT_PROVIDER_ID = 'default';

export function readAwsS3Configs(config: Config): AwsS3Config[] {
  const configs: AwsS3Config[] = [];

  const providerConfigs = config.getOptionalConfig('catalog.providers.awsS3');
  if (!providerConfigs) {
    return configs;
  }

  if (providerConfigs.has('bucketName')) {
    // simple/single config variant
    configs.push(readAwsS3Config(DEFAULT_PROVIDER_ID, providerConfigs));

    return configs;
  }

  for (const id of providerConfigs.keys()) {
    configs.push(readAwsS3Config(id, providerConfigs.getConfig(id)));
  }

  return configs;
}

function readAwsS3Config(id: string, config: Config): AwsS3Config {
  const bucketName = config.getString('bucketName');
  const region = config.getOptionalString('region');
  const prefix = config.getOptionalString('prefix');
  const accountId = config.getOptionalString('accountId');

  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;

  return {
    id,
    bucketName,
    region,
    prefix,
    schedule,
    accountId,
  };
}
