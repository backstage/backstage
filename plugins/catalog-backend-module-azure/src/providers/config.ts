/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { readSchedulerServiceTaskScheduleDefinitionFromConfig } from '@backstage/backend-plugin-api';
import { Config } from '@backstage/config';
import { AzureDevOpsConfig, AzureBlobStorageConfig } from './types';

const DEFAULT_PROVIDER_ID = 'default';

export function readAzureDevOpsConfigs(config: Config): AzureDevOpsConfig[] {
  const configs: AzureDevOpsConfig[] = [];

  const providerConfigs = config.getOptionalConfig(
    'catalog.providers.azureDevOps',
  );

  if (!providerConfigs) {
    return configs;
  }

  for (const id of providerConfigs.keys()) {
    configs.push(readAzureDevOpsConfig(id, providerConfigs.getConfig(id)));
  }

  return configs;
}

function readAzureDevOpsConfig(id: string, config: Config): AzureDevOpsConfig {
  const organization = config.getString('organization');
  const project = config.getString('project');
  const host = config.getOptionalString('host') || 'dev.azure.com';
  const repository = config.getOptionalString('repository') || '*';
  const branch = config.getOptionalString('branch');
  const path = config.getOptionalString('path') || '/catalog-info.yaml';

  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;

  return {
    id,
    host,
    organization,
    project,
    repository,
    branch,
    path,
    schedule,
  };
}

export function readAzureBlobStorageConfigs(
  config: Config,
): AzureBlobStorageConfig[] {
  const configs: AzureBlobStorageConfig[] = [];

  const providerConfigs = config.getOptionalConfig(
    'catalog.providers.azureBlob',
  );

  if (!providerConfigs) {
    return configs;
  }

  if (providerConfigs.has('containerName')) {
    // simple/single config variant
    configs.push(
      readAzureBlobStorageConfig(DEFAULT_PROVIDER_ID, providerConfigs),
    );

    return configs;
  }

  for (const id of providerConfigs.keys()) {
    configs.push(readAzureBlobStorageConfig(id, providerConfigs.getConfig(id)));
  }

  return configs;
}

function readAzureBlobStorageConfig(
  id: string,
  config: Config,
): AzureBlobStorageConfig {
  const containerName = config.getString('containerName');
  const accountName = config.getString('accountName');
  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;

  return {
    id,
    containerName,
    accountName,
    schedule,
  };
}
