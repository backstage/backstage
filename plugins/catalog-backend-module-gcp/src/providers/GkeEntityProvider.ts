/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  DeferredEntity,
  EntityProvider,
  EntityProviderConnection,
} from '@backstage/plugin-catalog-node';
import * as container from '@google-cloud/container';
import {
  ANNOTATION_KUBERNETES_API_SERVER,
  ANNOTATION_KUBERNETES_API_SERVER_CA,
  ANNOTATION_KUBERNETES_AUTH_PROVIDER,
  ANNOTATION_KUBERNETES_DASHBOARD_APP,
  ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS,
} from '@backstage/plugin-kubernetes-common';
import { Config } from '@backstage/config';
import {
  LoggerService,
  SchedulerService,
  SchedulerServiceTaskRunner,
  readSchedulerServiceTaskScheduleDefinitionFromConfig,
} from '@backstage/backend-plugin-api';
import {
  ANNOTATION_LOCATION,
  ANNOTATION_ORIGIN_LOCATION,
} from '@backstage/catalog-model';

/**
 * Catalog provider to ingest GKE clusters
 *
 * @public
 */
export class GkeEntityProvider implements EntityProvider {
  private readonly logger: LoggerService;
  private readonly scheduleFn: () => Promise<void>;
  private readonly gkeParents: string[];
  private readonly clusterManagerClient: container.v1.ClusterManagerClient;
  private connection?: EntityProviderConnection;

  private constructor(
    logger: LoggerService,
    taskRunner: SchedulerServiceTaskRunner,
    gkeParents: string[],
    clusterManagerClient: container.v1.ClusterManagerClient,
  ) {
    this.logger = logger;
    this.scheduleFn = this.createScheduleFn(taskRunner);
    this.gkeParents = gkeParents;
    this.clusterManagerClient = clusterManagerClient;
  }

  public static fromConfig({
    logger,
    scheduler,
    config,
  }: {
    logger: LoggerService;
    scheduler: SchedulerService;
    config: Config;
  }) {
    return GkeEntityProvider.fromConfigWithClient({
      logger,
      scheduler: scheduler,
      config,
      clusterManagerClient: new container.v1.ClusterManagerClient(),
    });
  }

  public static fromConfigWithClient({
    logger,
    scheduler,
    config,
    clusterManagerClient,
  }: {
    logger: LoggerService;
    scheduler: SchedulerService;
    config: Config;
    clusterManagerClient: container.v1.ClusterManagerClient;
  }) {
    const gkeProviderConfig = config.getConfig('catalog.providers.gcp.gke');
    const schedule = readSchedulerServiceTaskScheduleDefinitionFromConfig(
      gkeProviderConfig.getConfig('schedule'),
    );
    return new GkeEntityProvider(
      logger,
      scheduler.createScheduledTaskRunner(schedule),
      gkeProviderConfig.getStringArray('parents'),
      clusterManagerClient,
    );
  }

  getProviderName(): string {
    return `gcp-gke`;
  }

  async connect(connection: EntityProviderConnection): Promise<void> {
    this.connection = connection;
    await this.scheduleFn();
  }

  private filterOutUndefinedDeferredEntity(
    e: DeferredEntity | undefined,
  ): e is DeferredEntity {
    return e !== undefined;
  }

  private filterOutUndefinedCluster(
    c: container.protos.google.container.v1.ICluster | null | undefined,
  ): c is container.protos.google.container.v1.ICluster {
    return c !== undefined && c !== null;
  }

  private clusterToResource(
    cluster: container.protos.google.container.v1.ICluster,
    project: string,
  ): DeferredEntity | undefined {
    const location = `${this.getProviderName()}:${cluster.location}`;

    if (
      !cluster.name ||
      !cluster.selfLink ||
      !cluster.endpoint ||
      !cluster.location
    ) {
      this.logger.warn(
        `ignoring partial cluster, one of name=${cluster.name}, endpoint=${cluster.endpoint}, selfLink=${cluster.selfLink} or location=${cluster.location} is missing`,
      );
      return undefined;
    }

    // TODO fix location type
    return {
      locationKey: location,
      entity: {
        apiVersion: 'backstage.io/v1alpha1',
        kind: 'Resource',
        metadata: {
          annotations: {
            [ANNOTATION_KUBERNETES_API_SERVER]: `https://${cluster.endpoint}`,
            [ANNOTATION_KUBERNETES_API_SERVER_CA]:
              cluster.masterAuth?.clusterCaCertificate || '',
            [ANNOTATION_KUBERNETES_AUTH_PROVIDER]: 'google',
            [ANNOTATION_KUBERNETES_DASHBOARD_APP]: 'gke',
            [ANNOTATION_LOCATION]: location,
            [ANNOTATION_ORIGIN_LOCATION]: location,
            [ANNOTATION_KUBERNETES_DASHBOARD_PARAMETERS]: JSON.stringify({
              projectId: project,
              region: cluster.location,
              clusterName: cluster.name,
            }),
          },
          name: cluster.name,
          namespace: 'default',
        },
        spec: {
          type: 'kubernetes-cluster',
          owner: 'unknown',
        },
      },
    };
  }

  private createScheduleFn(
    taskRunner: SchedulerServiceTaskRunner,
  ): () => Promise<void> {
    return async () => {
      const taskId = `${this.getProviderName()}:refresh`;
      return taskRunner.run({
        id: taskId,
        fn: async () => {
          try {
            await this.refresh();
          } catch (error) {
            this.logger.error(error);
          }
        },
      });
    };
  }

  private async getClusters(): Promise<DeferredEntity[]> {
    const clusters = await Promise.all(
      this.gkeParents.map(async parent => {
        const project = parent.split('/')[1];
        const request = {
          parent: parent,
        };
        const [response] = await this.clusterManagerClient.listClusters(
          request,
        );
        return (
          response.clusters
            ?.filter(this.filterOutUndefinedCluster)
            .map(c => this.clusterToResource(c, project))
            .filter(this.filterOutUndefinedDeferredEntity) ?? []
        );
      }),
    );
    return clusters.flat();
  }

  async refresh() {
    if (!this.connection) {
      throw new Error('Not initialized');
    }

    this.logger.info('Discovering GKE clusters');

    let resources: DeferredEntity[];

    try {
      resources = await this.getClusters();
    } catch (e) {
      this.logger.error('error fetching GKE clusters', e);
      return;
    }

    this.logger.info(
      `Ingesting GKE clusters [${resources
        .map(r => r.entity.metadata.name)
        .join(', ')}]`,
    );

    await this.connection.applyMutation({
      type: 'full',
      entities: resources,
    });
  }
}
