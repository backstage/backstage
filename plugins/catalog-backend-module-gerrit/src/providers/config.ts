/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { readSchedulerServiceTaskScheduleDefinitionFromConfig } from '@backstage/backend-plugin-api';
import { Config } from '@backstage/config';
import { GerritProviderConfig } from './types';

function readGerritConfig(id: string, config: Config): GerritProviderConfig {
  const branch = config.getOptionalString('branch') ?? 'master';
  const host = config.getString('host');
  const query = config.getString('query');

  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;

  return {
    branch,
    host,
    id,
    query,
    schedule,
  };
}

export function readGerritConfigs(config: Config): GerritProviderConfig[] {
  const configs: GerritProviderConfig[] = [];

  const providerConfigs = config.getOptionalConfig('catalog.providers.gerrit');

  if (!providerConfigs) {
    return configs;
  }

  for (const id of providerConfigs.keys()) {
    configs.push(readGerritConfig(id, providerConfigs.getConfig(id)));
  }

  return configs;
}
