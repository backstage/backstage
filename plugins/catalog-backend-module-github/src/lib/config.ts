/*
 * Copyright 2020 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Config } from '@backstage/config';

/**
 * The configuration parameters for a multi-org GitHub processor.
 * @public
 */
export type GithubMultiOrgConfig = Array<{
  /**
   * The name of the GitHub org to process.
   */
  name: string;
  /**
   * The namespace of the group created for this org.
   */
  groupNamespace: string;
  /**
   * The namespace of the users created for this org. If not specified defaults to undefined.
   */
  userNamespace: string | undefined;
}>;

export function readGithubMultiOrgConfig(config: Config): GithubMultiOrgConfig {
  const orgConfigs = config.getOptionalConfigArray('orgs') ?? [];
  return orgConfigs.map(c => ({
    name: c.getString('name'),
    groupNamespace: (
      c.getOptionalString('groupNamespace') ?? c.getString('name')
    ).toLowerCase(),
    userNamespace: c.getOptionalString('userNamespace') ?? undefined,
  }));
}
