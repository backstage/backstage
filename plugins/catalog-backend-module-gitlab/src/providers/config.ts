/*
 * Copyright 2022 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { readSchedulerServiceTaskScheduleDefinitionFromConfig } from '@backstage/backend-plugin-api';
import { Config } from '@backstage/config';
import { GitlabProviderConfig } from '../lib';

/**
 * Extracts the gitlab config from a config object
 *
 * @public
 *
 * @param id - The provider key
 * @param config - The config object to extract from
 */
function readGitlabConfig(id: string, config: Config): GitlabProviderConfig {
  const group = config.getOptionalString('group') ?? '';
  const host = config.getString('host');
  const branch = config.getOptionalString('branch');
  const fallbackBranch = config.getOptionalString('fallbackBranch') ?? 'master';
  const catalogFile =
    config.getOptionalString('entityFilename') ?? 'catalog-info.yaml';
  const projectPattern = new RegExp(
    config.getOptionalString('projectPattern') ?? /[\s\S]*/,
  );
  const userPattern = new RegExp(
    config.getOptionalString('userPattern') ?? /[\s\S]*/,
  );
  const groupPattern = new RegExp(
    config.getOptionalString('groupPattern') ?? /[\s\S]*/,
  );
  const orgEnabled: boolean = config.getOptionalBoolean('orgEnabled') ?? false;
  const allowInherited: boolean =
    config.getOptionalBoolean('allowInherited') ?? false;
  const relations: string[] = config.getOptionalStringArray('relations') ?? [];

  const skipForkedRepos: boolean =
    config.getOptionalBoolean('skipForkedRepos') ?? false;
  const excludeRepos: string[] =
    config.getOptionalStringArray('excludeRepos') ?? [];

  const schedule = config.has('schedule')
    ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
        config.getConfig('schedule'),
      )
    : undefined;
  const restrictUsersToGroup =
    config.getOptionalBoolean('restrictUsersToGroup') ?? false;

  return {
    id,
    group,
    branch,
    fallbackBranch,
    host,
    catalogFile,
    projectPattern,
    userPattern,
    groupPattern,
    schedule,
    orgEnabled,
    allowInherited,
    relations,
    skipForkedRepos,
    excludeRepos,
    restrictUsersToGroup,
  };
}

/**
 * Extracts the gitlab config from a config object array
 *
 * @public
 *
 * @param config - The config object to extract from
 */
export function readGitlabConfigs(config: Config): GitlabProviderConfig[] {
  const configs: GitlabProviderConfig[] = [];

  const providerConfigs = config.getOptionalConfig('catalog.providers.gitlab');

  if (!providerConfigs) {
    return configs;
  }

  for (const id of providerConfigs.keys()) {
    configs.push(readGitlabConfig(id, providerConfigs.getConfig(id)));
  }

  return configs;
}
