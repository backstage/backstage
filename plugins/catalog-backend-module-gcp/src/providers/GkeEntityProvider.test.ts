/*
 * Copyright 2023 The Backstage Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { GkeEntityProvider } from './GkeEntityProvider';
import { SchedulerServiceTaskRunner } from '@backstage/backend-plugin-api';
import * as container from '@google-cloud/container';
import { ConfigReader } from '@backstage/config';

describe('GkeEntityProvider', () => {
  const clusterManagerClientMock = {
    listClusters: jest.fn(),
  };
  const connectionMock = {
    applyMutation: jest.fn(),
    refresh: jest.fn(),
  };
  const taskRunner = {
    createScheduleFn: jest.fn(),
    run: jest.fn(),
  } as SchedulerServiceTaskRunner;
  const schedulerMock = {
    createScheduledTaskRunner: jest.fn(),
  } as any;
  const logger = {
    info: jest.fn(),
    error: jest.fn(),
  };
  let gkeEntityProvider: GkeEntityProvider;

  beforeEach(async () => {
    jest.resetAllMocks();
    schedulerMock.createScheduledTaskRunner.mockReturnValue(taskRunner);
    gkeEntityProvider = GkeEntityProvider.fromConfigWithClient({
      logger: logger as any,
      config: new ConfigReader({
        catalog: {
          providers: {
            gcp: {
              gke: {
                parents: [
                  'projects/parent1/locations/-',
                  'projects/parent2/locations/some-other-location',
                ],
                schedule: {
                  frequency: {
                    minutes: 3,
                  },
                  timeout: {
                    minutes: 3,
                  },
                },
              },
            },
          },
        },
      }),
      scheduler: schedulerMock,
      clusterManagerClient: clusterManagerClientMock as any,
    });
    await gkeEntityProvider.connect(connectionMock);
  });

  it('should return clusters as Resources', async () => {
    clusterManagerClientMock.listClusters.mockImplementation(req => {
      if (req.parent === 'projects/parent1/locations/-') {
        return [
          {
            clusters: [
              {
                name: 'some-cluster',
                endpoint: '127.0.0.1',
                location: 'some-location',
                selfLink: 'https://127.0.0.1/some-link',
                masterAuth: {
                  clusterCaCertificate: 'abcdefg',
                },
              },
            ],
          },
        ];
      } else if (
        req.parent === 'projects/parent2/locations/some-other-location'
      ) {
        return [
          {
            clusters: [
              {
                name: 'some-other-cluster',
                endpoint: '127.0.0.1',
                location: 'some-other-location',
                selfLink: 'https://127.0.0.1/some-other-link',
                masterAuth: {
                  // no CA cert is ok
                },
              },
            ],
          },
        ];
      }

      throw new Error(`unexpected parent ${req.parent}`);
    });
    await gkeEntityProvider.refresh();
    expect(connectionMock.applyMutation).toMatchSnapshot();
  });

  const ignoredPartialClustersTests: [
    string,
    container.protos.google.container.v1.ICluster,
  ][] = [
    [
      'no-cluster-name',
      {
        endpoint: '127.0.0.1',
        location: 'some-location',
        selfLink: 'http://127.0.0.1/some-link',
        masterAuth: {
          clusterCaCertificate: 'abcdefg',
        },
      },
    ],
    [
      'no-self-link',
      {
        // no selfLink
        name: 'some-name',
        endpoint: '127.0.0.1',
        location: 'some-location',
        masterAuth: {
          clusterCaCertificate: 'abcdefg',
        },
      },
    ],
    [
      'no-endpoint',
      {
        name: 'some-name',
        location: 'some-location',
        selfLink: 'http://127.0.0.1/some-link',
        masterAuth: {
          clusterCaCertificate: 'abcdefg',
        },
      },
    ],
    [
      'no-location',
      {
        name: 'some-name',
        endpoint: '127.0.0.1',
        selfLink: 'http://127.0.0.1/some-link',
        masterAuth: {
          clusterCaCertificate: 'abcdefg',
        },
      },
    ],
  ];

  it.each(ignoredPartialClustersTests)(
    'ignore cluster - %s',
    async (_name, ignoredCluster) => {
      clusterManagerClientMock.listClusters.mockImplementation(req => {
        if (req.parent === 'projects/parent1/locations/-') {
          return [ignoredCluster];
        }
        return [
          {
            clusters: [],
          },
        ];
      });
      await gkeEntityProvider.refresh();
      expect(connectionMock.applyMutation).toHaveBeenCalledWith({
        type: 'full',
        entities: [],
      });
    },
  );

  it('should log GKE API errors', async () => {
    clusterManagerClientMock.listClusters.mockRejectedValue(
      new Error('some-error'),
    );
    await gkeEntityProvider.refresh();
    expect(connectionMock.applyMutation).toHaveBeenCalledTimes(0);
    expect(logger.error).toHaveBeenCalledTimes(1);
  });
});
