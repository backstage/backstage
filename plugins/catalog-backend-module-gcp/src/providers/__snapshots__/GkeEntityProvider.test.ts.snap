// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GkeEntityProvider should return clusters as Resources 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "entities": [
          {
            "entity": {
              "apiVersion": "backstage.io/v1alpha1",
              "kind": "Resource",
              "metadata": {
                "annotations": {
                  "backstage.io/managed-by-location": "gcp-gke:some-location",
                  "backstage.io/managed-by-origin-location": "gcp-gke:some-location",
                  "kubernetes.io/api-server": "https://127.0.0.1",
                  "kubernetes.io/api-server-certificate-authority": "abcdefg",
                  "kubernetes.io/auth-provider": "google",
                  "kubernetes.io/dashboard-app": "gke",
                  "kubernetes.io/dashboard-parameters": "{"projectId":"parent1","region":"some-location","clusterName":"some-cluster"}",
                },
                "name": "some-cluster",
                "namespace": "default",
              },
              "spec": {
                "owner": "unknown",
                "type": "kubernetes-cluster",
              },
            },
            "locationKey": "gcp-gke:some-location",
          },
          {
            "entity": {
              "apiVersion": "backstage.io/v1alpha1",
              "kind": "Resource",
              "metadata": {
                "annotations": {
                  "backstage.io/managed-by-location": "gcp-gke:some-other-location",
                  "backstage.io/managed-by-origin-location": "gcp-gke:some-other-location",
                  "kubernetes.io/api-server": "https://127.0.0.1",
                  "kubernetes.io/api-server-certificate-authority": "",
                  "kubernetes.io/auth-provider": "google",
                  "kubernetes.io/dashboard-app": "gke",
                  "kubernetes.io/dashboard-parameters": "{"projectId":"parent2","region":"some-other-location","clusterName":"some-other-cluster"}",
                },
                "name": "some-other-cluster",
                "namespace": "default",
              },
              "spec": {
                "owner": "unknown",
                "type": "kubernetes-cluster",
              },
            },
            "locationKey": "gcp-gke:some-other-location",
          },
        ],
        "type": "full",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
